uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/LeaveRequest.php

```diff
@@ -768,10 +780,9 @@ private function saveDates() {
   }
 
   /**
-   * Deletes all the dates and balance changes related to this LeaveRequest
+   * Deletes all the dates related to this LeaveRequest
    */
-  private function deleteDatesAndBalanceChanges() {
-    LeaveBalanceChange::deleteAllForLeaveRequest($this);
+  private function deleteDates() {
     LeaveRequestDate::deleteDatesForLeaveRequest($this->id);
```

Comment:
Can you explain why we're not deleting the balance changes here anymore? If we delete the dates but not their balance changes, we'll end up with a lot of orphan records in the database

Reply:
I moved the function to https://github.com/civicrm/civihr/blob/ca0b8ee912f946e2186d8426469f6a56e864594b/uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Service/LeaveBalanceChange.php#L15-L15.
Think of this scenario: The user modifies the status of an already created leave request and change_balance = 0, in the leave request service, we don't recalculate balance change but meanwhile during the leave request update in the BAO, the balance change would have been deleted here and the user ends up with Zero balance change for the request.

Reply:
In that scenario, the dates won't have changed, which means the `deleteDates` (or `deleteDatesAndBalanceChanges`) method won't be called and the balance changes shouldn't be touched as well.

Calling `LeaveBalanceChange::deleteAllForLeaveRequest` AFTER the leave request dates have been deleted will simply not delete anything, because the balance changes are linked to the request dates, not to the leave request itself. If you delete the dates first, you delete the link between the leave request and the balance changes and there's no way to know which balance changes belong to the request.

Reply:
You're right. Thanks. Will revert. Wondering what I was thinking. (shrug)

Reply:
Just thought of another scenario why I added it now. A scenario where the user modifies an already created leave request without changing the dates and decides to change the balance change i.e change_balance = 1. In this case, the dates and balance changes will not be deleted and a new balance change will be added to the already existing one causing an increase in balance change. For this case, calling `LeaveBalanceChange::deleteAllForLeaveRequest($leaveRequest);` will successfully delete the old leave requests balances because the dates will still exist and prevent the bug.

Reply:
Not sure what to do. Maybe I should revert to delete leave request dates and balance changes so we don't have orphan records in the case when the date changes in addition to adding the method in  https://github.com/civicrm/civihr/blob/ca0b8ee912f946e2186d8426469f6a56e864594b/uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Service/LeaveBalanceChange.php#L15-L15

Reply:
Yeah, sounds like a good plan